### PR TITLE
🌱 MHC should not remediate paused machines

### LINKED ISSUE #3421: MHC should respect pause annotation on Machines
Today, MHC applies remediation or health checks regardless if a Machine has been paused or not. We should change this behavior and check if the Machine is paused before performing any action.

The main use case this will allow is to avoid deleting infrastructure if cluster admins are performing operations on the nodes.

/kind cleanup
/milestone v0.3.x

### PR DESCRIPTION
**Pull request by benmoss**
**What this PR does / why we need it**:
Right now the MHC controller only checks if the MHC object itself or the Cluster is paused. It seems reasonable users would expect that pausing a machine would also pause remediation on it.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #3421

In #3108 I unintentionally changed the logic here to return when we encounter an error patching one of the machines, instead of the prior behavior of continuing to try to remediate all of the machines before returning an error collection. This also brings back that logic.

### CODE REVIEW

File: controllers/machinehealthcheck_controller.go [hunk 1]
```diff
@@ -250,12 +250,17 @@ func (r *MachineHealthCheckReconciler) reconcile(ctx context.Context, logger log
...
+			continue
+		}
 		condition := conditions.Get(t.Machine, clusterv1.MachineHealthCheckSuccededCondition)
 		logger.Info("Target has failed health check, marking for remediation", "target", t.string(), "reason", condition.Reason, "message", condition.Message)
 
 		conditions.MarkFalse(t.Machine, clusterv1.MachineOwnerRemediatedCondition, clusterv1.WaitingForRemediation, clusterv1.ConditionSeverityWarning, "MachineHealthCheck failed")
 		if err := t.patchHelper.Patch(ctx, t.Machine); err != nil {
-			return ctrl.Result{}, errors.Wrapf(err, "Failed to patch unhealthy machine status for machine %q", t.Machine.Name)
+			errList = append(errList, errors.Wrapf(err, "Failed to patch unhealthy machine status for machine %q", t.Machine.Name))
+			continue
```
Comment (vincepri) [inline controllers/machinehealthcheck_controller.go:263]: Is this change not related to this PR?

File: controllers/machinehealthcheck_controller.go [hunk 2]
```diff
@@ -267,7 +272,8 @@ func (r *MachineHealthCheckReconciler) reconcile(ctx context.Context, logger log
 	}
 	for _, t := range healthy {
 		if err := t.patchHelper.Patch(ctx, t.Machine); err != nil {
-			return ctrl.Result{}, errors.Wrapf(err, "Failed to patch healthy machine status for machine %q", t.Machine.Name)
+			errList = append(errList, errors.Wrapf(err, "Failed to patch healthy machine status for machine %q", t.Machine.Name))
+			continue
```
Comment (vincepri) [inline controllers/machinehealthcheck_controller.go:276]: Is this change not related to this PR?

File: controllers/machinehealthcheck_controller.go [hunk 3]
```diff
@@ -250,12 +250,17 @@ func (r *MachineHealthCheckReconciler) reconcile(ctx context.Context, logger log
 	// mark for remediation
 	errList := []error{}
 	for _, t := range unhealthy {
+		if annotations.HasPausedAnnotation(t.Machine) {
```
Comment (detiber) [inline controllers/machinehealthcheck_controller.go:253]: This could probably use a comment as to why it's not using `annotations.IsPaused` (primarily because it's called in `Reconcile` and if cluster.Spec.Paused was true we wouldn't make it here).

File: controllers/machinehealthcheck_controller.go (hunk 1 above)
Comment (benmoss) [reply controllers/machinehealthcheck_controller.go:263]: It's related since it's where I noticed the problem of the unused `errList` and realized I didn't intentionally mean to change the logic of MHC to not try to patch each machine and instead return early after the first error.

File: controllers/machinehealthcheck_controller.go (hunk 2 above)
Comment (benmoss) [reply controllers/machinehealthcheck_controller.go:276]: It's related since it's where I noticed the problem of the unused `errList` and realized I didn't intentionally mean to change the logic of MHC to not try to patch each machine and instead return early after the first error.

File: controllers/machinehealthcheck_controller.go (hunk 3 above)
Comment (benmoss) [reply controllers/machinehealthcheck_controller.go]: I just switched it to use `IsPaused`, that seems simpler than adding a comment

**Comment by detiber** [comment]
changes lgtm pending @vincepri's request for a test

**Comment by benmoss** [comment]
Do we actually want to mark it with `clusterv1.MachineHealthCheckSucceededCondition = false` and just not put the remediation needed condition on?

**Comment by benmoss** [comment]
I think that's the right thing, so I've updated the implementation to do so

File: controllers/machinehealthcheck_controller.go (hunk 1 above)
Comment (vincepri) [reply controllers/machinehealthcheck_controller.go]: Got it, thanks for explaining

File: controllers/machinehealthcheck_controller.go [hunk 4]
```diff
@@ -251,9 +251,13 @@ func (r *MachineHealthCheckReconciler) reconcile(ctx context.Context, logger log
 	errList := []error{}
 	for _, t := range unhealthy {
 		condition := conditions.Get(t.Machine, clusterv1.MachineHealthCheckSuccededCondition)
-		logger.Info("Target has failed health check, marking for remediation", "target", t.string(), "reason", condition.Reason, "message", condition.Message)
 
-		conditions.MarkFalse(t.Machine, clusterv1.MachineOwnerRemediatedCondition, clusterv1.WaitingForRemediation, clusterv1.ConditionSeverityWarning, "MachineHealthCheck failed")
+		if annotations.IsPaused(cluster, t.Machine) {
+			logger.Info("Machine has failed health check, but machine is paused so skipping remediation", "target", t.string(), "reason", condition.Reason, "message", condition.Message)
```
Comment (fabriziopandini) [inline controllers/machinehealthcheck_controller.go:256]: My only concern is that the user could be confused by a machine being unhealthy and remediation requests not being generated (with the reason explained in logs only). 
Might be we should consider adding a note in the in unhealthy condition message, wdyt?

Comment (benmoss) [reply controllers/machinehealthcheck_controller.go:256]: Hmm, do you think that a user wouldn't know that they paused their machine?

Comment (benmoss) [reply controllers/machinehealthcheck_controller.go:256]: Maybe paused should be a condition too?

Comment (vincepri) [reply controllers/machinehealthcheck_controller.go:256]: Let's open a different issue to add a condition for paused in v0.4, for now this seems fine

Comment (fabriziopandini) [reply controllers/machinehealthcheck_controller.go:256]: I was more thinking to a whole cluster paused, but I'm ok in tackling UX concern in a separated issue

Comment (benmoss) [reply controllers/machinehealthcheck_controller.go:256]: If the whole cluster is paused it will not reach this code because it is being handled [here](https://github.com/kubernetes-sigs/cluster-api/blob/3900bb1cb2cb77c6ceebeb747647b22db2817dba/controllers/machinehealthcheck_controller.go#L139-L143)

Comment (benmoss) [reply controllers/machinehealthcheck_controller.go:256]: https://github.com/kubernetes-sigs/cluster-api/issues/3450
